lib/thread/tests/mocks/freertos/semphr.h

```diff
@@ -4,35 +4,92 @@
 #ifndef FREERTOS_SEMPHR_MOCK_H_
 #define FREERTOS_SEMPHR_MOCK_H_
 
-#include <semaphore.h>
 #include <ecl/assert.h>
-#include <time.h>
+#include <condition_variable>
+#include <atomic>
 
-using SemaphoreHandle_t = sem_t *;
+// Implementation taken from POSIX semaphores.
 
-static inline SemaphoreHandle_t xSemaphoreCreateBinary()
+class helper_semphr
 {
-    auto *sem = new sem_t;
-    if (sem_init(sem, false, 0) < 0) {
-        ecl_assert_msg(0, "Cannot init semaphore");
+public:
+    helper_semphr()
+        :m_mutex{}
+        ,m_cond{}
+        ,m_flag{false}
+    {
+    }
+
+    void signal()
+    {
+        std::unique_lock<std::mutex> lock(m_mutex);
+        m_flag = true;
+        lock.unlock();
+
+        m_cond.notify_one();
+    }
+
+    bool try_wait(std::chrono::milliseconds ms)
+    {
+        std::unique_lock<std::mutex> lock(m_mutex);
+
+        // Already rised
+        if (m_flag) {
+            m_flag = false;
+            return true;
+        }
+
+        // Give a chance to raise flag
+        m_cond.wait_for(lock, ms, [&] { return m_flag.load(); });
```

Comment:
I consider using a return value from the `wait_for` to be a more precise way of detecting timeout. Otherwise, the flag is checked even if the `wait_for` was timed out.